File: artifact_one/original_code/Driver.java
/*
CS499 Example code for code review
Course IT-145
I was tasked with developing a software application 
in Java for the fictitious company Grazioso Salvare, 
which helps track search and rescue animals. 
These search and rescue animals are obtained and trained 
by the company for the purpose of rescuing humans from 
life-threatening situations.
Programmer: Konstantin Dobikov
Last Date updated: 06/03/2023
Artifact 1 Category 1 in Software Design and Engineering
No known errors
Enhancement Plan:
1.	Migrate the application to Python by porting Dog, Driver, Monkey, 
	and RescueAnimal classes in addition to Database, Users, and RescueAnimalDatabaseServices, 
	UsersDatabaseServices new classes.
2.	Add SQLite database for persistent storage
3.	Add login functionality
4.	Implement secure storage of hashed credentials.

*/

package grazioso;

import java.util.ArrayList;
import java.util.Scanner;

public class Driver {
    private static ArrayList<Dog> dogList = new ArrayList<Dog>();
    // Instance variables (if needed)
    private static ArrayList<Monkey> monkeyList = new ArrayList<Monkey>();
    //pattern for digits only
    private static final String DIGIT_PATTERN = "-?\\d+(\\.\\d+)?";
    //pattern for date format only
    private static final String DATE_PATTERN =
            "^(0[1-9]|[12][0-9]|3[01])/(0[1-9]|1[012])/((19|2[0-9])[0-9]{2})$";
    
    public static void main(String[] args) {
    	
    	Scanner scnr = new Scanner(System.in);
    	
        initializeDogList();
        initializeMonkeyList();
        
        // Loop that displays the menu, accepts the users input
        // and takes the appropriate action.
	
        String inputText;
        displayMenu();
        do {
        	inputText = scnr.nextLine();
            switch (inputText) {
            case "1":
            	intakeNewDog(scnr);
                break;
            case "2":
            	intakeNewMonkey(scnr);
                break;
            case "3":
            	reserveAnimal(scnr);
                break;
            case "4": 
            	printAnimals("dog");
                break;
            case "5":
            	printAnimals("monkey");
                break;
            case "6":
            	printAnimals("");
                break;
            case "q":
            	System.out.println("\nSee you later! Oops, saving data module is disabled!");
            	return;
            default:
            	System.out.println("Please use numbers from 1 to 6, or q to exit");
                break;
            }
        	displayMenu();
        } while(true);
    }

    // This method prints the menu options
    public static void displayMenu() {
        System.out.println("\n\n");
        System.out.println("\t\t\t\tRescue Animal System Menu");
        System.out.println("[1] Intake a new dog");
        System.out.println("[2] Intake a new monkey");
        System.out.println("[3] Reserve an animal");
        System.out.println("[4] Print a list of all dogs");
        System.out.println("[5] Print a list of all monkeys");
        System.out.println("[6] Print a list of all animals that are not reserved");
        System.out.println("[q] Quit application");
        System.out.println();
        System.out.println("Enter a menu selection");
    }


    // Adds dogs to a list for testing
    public static void initializeDogList() {
        Dog dog1 = new Dog("Spot", "German Shepherd", "male", "1", "25.6", "05-12-2019", "United States", "intake", false, "United States");
        Dog dog2 = new Dog("Rex", "Great Dane", "male", "3", "35.2", "02-03-2020", "United States", "Phase I", false, "United States");
        Dog dog3 = new Dog("Bella", "Chihuahua", "female", "4", "25.6", "12-12-2019", "Canada", "in-service", false, "Canada");
        Dog dog4 = new Dog("BellaClone", "Chihuahua", "female", "4", "25.6", "12-12-2019", "Canada", "in-service", false, "Canada");

        dogList.add(dog1);
        dogList.add(dog2);
        dogList.add(dog3);
        dogList.add(dog4);
        
    }


    // Adds monkeys to a list for testing
    //Optional for testing

    public static void initializeMonkeyList() {
        Monkey monkey1 = new Monkey("Coin", "Capuchin", "male", "1", "25.6", "05-12-2019", "United States", "intake", false, "United States", "2.0", "1.5", "3.0");
        Monkey monkey2 = new Monkey("Eth", "Guenon", "male", "3", "35.2", "02-03-2020", "United States", "Phase I", false, "United States", "3.0", "2.5", "4.0");
        Monkey monkey3 = new Monkey("Bitcoin", "Macaque", "female", "4", "25.6", "12-12-2019", "Canada", "in-service", false, "Canada", "4.0", "3.5", "5.0");

        monkeyList.add(monkey1);
        monkeyList.add(monkey2);
        monkeyList.add(monkey3);
    }


    // intakeNewDog method
    // Instantiate and add the new dog to the appropriate list with input validation
    
    public static void intakeNewDog(Scanner scanner) {
        System.out.println("What is the dog's name?");
        String name = scanner.nextLine();
        for(Dog dog: dogList) {
            if(dog.getName().equalsIgnoreCase(name)) {
                System.out.println("\n\nThis dog is already in our system\n\n");
                return; //returns to menu
            }
        }

        // Add the code to instantiate a new dog and add it to the appropriate list
        System.out.println("What is the dog's breed?");
        String breed = scanner.nextLine();
        System.out.println("What is the dog's gender?");
        String gender = scanner.nextLine();
        System.out.println("What is the dog's age?");
        String age = checkAnswer(scanner, DIGIT_PATTERN);
        System.out.println("What is the dog's weight in lbs?");
        String weight = checkAnswer(scanner, DIGIT_PATTERN);
        System.out.println("What is the dog's acquisition date? In format: dd/mm/yyyy");
        String acquisitionDate = checkAnswer(scanner, DATE_PATTERN);
        System.out.println("What is the dog's acquisition country?");
        String acquisitionCountry = scanner.nextLine();
        System.out.println("What is the dog's training status?\n"
        		+ "Possible answers: intake, farm, in-service, farm, Phase (I, II, III, IV, V)");
        // getting training status of a dog
        String trainingStatus = scanner.nextLine();
        boolean choice; //need it to stop the loop when we get the correct training status 
        do {
        	choice = true;
        	if (!trainingStatus.equals("intake") 
        			&& !trainingStatus.equals("in-service") 
        			&& !trainingStatus.equals("farm")
        			&& !trainingStatus.equals("Phase I")
        			&& !trainingStatus.equals("Phase II")
        			&& !trainingStatus.equals("Phase III")
        			&& !trainingStatus.equals("Phase IV")
        			&& !trainingStatus.equals("Phase V")) {
        		System.out.println("Whrong status! Please use the correct status!");
        		trainingStatus = scanner.nextLine();
        	}
        	else {
        		System.out.println("Thank you!\nAdding the dog in our database...");
        		choice = false;
        	}
        	
        } while(choice);
        // default not reserved
        boolean reserved = false;
        // default acquisition country is equal to service country
        String inServiceCountry = acquisitionCountry;
        // create instance of Dog and add to the ArrayList dogList
        Dog dog = new Dog(name, breed, gender, age,
        	weight, acquisitionDate, acquisitionCountry,
        	trainingStatus, reserved, inServiceCountry);
        dogList.add(dog);
        System.out.println("Dog was added.\nReturning to the main menu...");
        return;
    }


        // intakeNewMonkey
		// Instantiate and add the new monkey to the appropriate list with input validation

        public static void intakeNewMonkey(Scanner scanner) {
        	System.out.println("What is the monkey's name?");
            String name = scanner.nextLine();
            for(Monkey monkey: monkeyList) {
                if(monkey.getName().equalsIgnoreCase(name)) {
                    System.out.println("\n\nThis monkey is already in our system\n\n");
                    return; //returns to menu
                }
            }
            System.out.println("What is the monkey's species?");
            String species = scanner.nextLine();
            if (!species.equalsIgnoreCase("Capuchin") 
            		&& !species.equalsIgnoreCase("Guenon") 
            		&& !species.equalsIgnoreCase("Macaque") 
            		&& !species.equalsIgnoreCase("Marmoset") 
            		&& !species.equalsIgnoreCase("Squirrel monkey") 
            		&& !species.equalsIgnoreCase("Tamarin")) {
                System.out.println("\n\nThis monkey species is not allowed\n\n");
                return; 
            }
            System.out.println("What is the monkey's gender?");
            String gender = scanner.nextLine();
           
            System.out.println("What is the monkey's age?");
            String age = checkAnswer(scanner, DIGIT_PATTERN);
            System.out.println("What is the monkey's weight in lbs?");
            String weight = checkAnswer(scanner, DIGIT_PATTERN);
            System.out.println("What is the monkey's acquisition date? In format: dd/mm/yyyy");
            String acquisitionDate = checkAnswer(scanner, DATE_PATTERN);
            System.out.println("What is the monkey's acquisition country?");
            String acquisitionCountry = scanner.nextLine();
            //default status is intake
            String trainingStatus = "intake";
            //default is not reserved
            boolean reserved = false;
            //default acquisition country is equal to service country
            String inServiceCountry = acquisitionCountry;
            //height, bodyLength, tailLength
            System.out.println("What is the monkey's height in inches?");
            String height = checkAnswer(scanner, DIGIT_PATTERN);
            
            System.out.println("What is the monkey's body length inches?");
            String bodyLength = checkAnswer(scanner, DIGIT_PATTERN);
            
            System.out.println("What is the monkey's tail length inches?");
            String tailLength = checkAnswer(scanner, DIGIT_PATTERN);

            //create instance of Monkey and add to the ArrayList monkeyList
            Monkey monkey = new Monkey(name, species, gender, age,
            	weight, acquisitionDate, acquisitionCountry,
            	trainingStatus, reserved, inServiceCountry,
            	height, bodyLength, tailLength);
            monkeyList.add(monkey);
            System.out.println("Monkey was added. Returning to the main menu.");
            return;
        }

        // reserveAnimal method
        // We will need to find the animal by animal type and in service country
        public static void reserveAnimal(Scanner scanner) {
        	System.out.println("Please enter animal's type to be reserved?");
            String type = scanner.nextLine();
            boolean choiceAnimal;
            do {
            	choiceAnimal = true;
            	if (!type.equalsIgnoreCase("dog") 
            			&& !type.equalsIgnoreCase("monkey")) {
            		System.out.println("Whrong type!\nPossible types: dog, monkey");
            		type = scanner.nextLine();
            	}
            	else {
            		choiceAnimal = false;
            	}
            	
            } while(choiceAnimal);
        	System.out.println("Please enter animal's country to be reserved?");
            String inServiceCountry = scanner.nextLine();
            if (type.equalsIgnoreCase("dog")) {
            	for(Dog dog: dogList) {
            		if (dog.getInServiceLocation().equalsIgnoreCase(inServiceCountry) 
            				&& dog.getTrainingStatus().equals("in-service") 
            				&& !dog.getReserved()) {
            				dog.setReserved(true);
            				System.out.println(dog.getName() + " is reserved!");
            				return;
            	    }
            	}
            	noAvail();
            } 
            else if (type.equalsIgnoreCase("monkey")) {
            	for(Monkey monkey: monkeyList) {
            		if (monkey.getInServiceLocation().equalsIgnoreCase(inServiceCountry) 
            				&& monkey.getTrainingStatus().equals("in-service") 
            				&& !monkey.getReserved()) {
            				monkey.setReserved(true);
            				System.out.println(monkey.getName() + " is reserved!");
            				return;
            		} 
            	}
            	noAvail();
            }
            else {
            	noAvail();
            }
        }
        public static void noAvail() {
        	System.out.println("No animal to reserve in this country! Come back later!");
        	return;
        }
        // test method to reserve a dog
        public static void dogReserve(ArrayList<Dog> dogList, String inServiceCountry) {
        	for(Dog dog: dogList) {
        		if (dog.getInServiceLocation().equals(inServiceCountry) && dog.getTrainingStatus().equals("in-service") && !dog.getReserved()) {
        				dog.setReserved(true);
        				System.out.println(dog.getName() + " is reserved!");
        				return;
        	    }
        	}
        	noAvail();
        }
        // test method to reserve a monkey
        public static void monkeyReserve(ArrayList<Monkey> monkeyList, String inServiceCountry) {
        	for(Monkey monkey: monkeyList) {
        		if (monkey.getInServiceLocation().equals(inServiceCountry) && monkey.getTrainingStatus().equals("in-service") && !monkey.getReserved()) {
        				monkey.setReserved(true);
        				System.out.println(monkey.getName() + " is reserved!");
        				return;
        		} 
        	}
        	noAvail();
        }
        // method to check to get the right answer
        public static String checkAnswer(Scanner scanner, String pattern) {
        	String answer;
        	boolean rightAnswer = true;
        	answer = scanner.nextLine();
        	do {
        		if(answer.matches(pattern)){
        			rightAnswer = false;
        	}
        	else {
        		if (pattern == DIGIT_PATTERN) {
        			System.out.print("Not allowed, please use numbers only: ");
        		} 
        		else if (pattern == DATE_PATTERN) {
        			System.out.print("Not allowed, please use the right format dd/mm/yyyy: ");
        		}
            	answer = scanner.nextLine();
             }
        	} while(rightAnswer);
        	return answer;
        }
        
        // method printAnimals
        // Includes the animal name, status, acquisition country and if the animal is reserved.
        // This method connects to three different menu items.
        // The printAnimals() method has three different outputs
        // based on the listType parameter
        // dog - prints the list of dogs
        // monkey - prints the list of monkeys
        // available - prints a combined list of all animals that are
        // fully trained ("in service") but not reserved 
        public static void printAnimals(String listType) {
        	if (listType.equals("monkey")) {
        		System.out.printf("-------------------------------------------------------------------\n");
        		System.out.printf("| %-10s | %18s | %16s | %10s |\n", "Name", "Training Status", "Location", "Reserved");
        		System.out.printf("-------------------------------------------------------------------\n");
        		for(Monkey monkey: monkeyList) {
        			System.out.printf("| %-10s | %18s | %16s | %10s |\n", monkey.getName(), monkey.getTrainingStatus(), monkey.getAcquisitionLocation(), monkey.getReserved());           
        		}
        		System.out.printf("-------------------------------------------------------------------\n");
        	}
        	else if (listType.equals("dog")) {
        		System.out.printf("-------------------------------------------------------------------\n");
        		System.out.printf("| %-10s | %18s | %16s | %10s |\n", "Name", "Training Status", "Location", "Reserved");
        		System.out.printf("-------------------------------------------------------------------\n");
        		for(Dog dog: dogList) {
        			System.out.printf("| %-10s | %18s | %16s | %10s |\n", dog.getName(), dog.getTrainingStatus(), dog.getAcquisitionLocation(), dog.getReserved());
        		}
        		System.out.printf("-------------------------------------------------------------------\n");
        	}        		
        	else {
                int i = 0; //count of reserved animals
                System.out.printf("--------------------------------------------------------------------------------\n");
        		System.out.printf("| %-10s | %10s | %18s | %16s | %10s |\n", "Name", "Type", "Training Status", "Location", "Reserved");
        		System.out.printf("--------------------------------------------------------------------------------\n");
                for(Dog dog: dogList) {
             	   if (dog.getTrainingStatus().equals("in-service") && !dog.getReserved()) {
             		  System.out.printf("| %-10s | %10s | %18s | %16s | %10s |\n", dog.getName(), "Dog", dog.getTrainingStatus(), dog.getAcquisitionLocation(), dog.getReserved());
                        ++i;
             	   }
            		}
                for(Monkey monkey: monkeyList) {
             	   if (monkey.getTrainingStatus().equals("in-service") && !monkey.getReserved()) {
             		  System.out.printf("| %-10s | %10s | %18s | %16s | %10s |\n", monkey.getName(), "Monkey", monkey.getTrainingStatus(), monkey.getAcquisitionLocation(), monkey.getReserved());
                        ++i;
             	   }
                }
                System.out.printf("--------------------------------------------------------------------------------\n");
                System.out.println();
                if (i == 0) {
             	   System.out.println("No animals available to reserve! Comeback later!");
                }
        	}
        }

}

File: artifact_one/enhanced_code/Driver.py
#import database servicies
import DataBaseServices
# time library to protect from bruteforcing
# username and password
import time
# regex library to filter user input
import re
# to hide the inputted password with maskpass()
import maskpass

# pattern for digits only
DIGIT_PATTERN = "-?\\d+(\\.\\d+)?"

# pattern for date format only
DATE_PATTERN = "^(0[1-9]|[12][0-9]|3[01])/(0[1-9]|1[012])/((19|2[0-9])[0-9]{2})$"

# List of possible trainingStatus
trainingStatusAllowList = ["intake", "farm", "in-service", "farm", "Phase I", "Phase II", "Phase III", "Phase IV", "Phase V"]

# List of allowed Monkeys
allowedMonkeysSpeciesList = ["Capuchin", "Guenon", "Macaque", "Marmoset", "Squirrel monkey", "Tamarin"]

class Driver:
    # default contructor
    def __init__(self, database_file):
        self.dataBaseServices = DataBaseServices.DataBaseServices(database_file)
    
    # method to check answer based on the given pattern
    def checkAnswer(self, pattern):
        rightAnswer = True
        answer = input()
        while(rightAnswer):
            if(re.search(pattern, answer)):
                rightAnswer = False
            else:
                if (pattern == DIGIT_PATTERN):
                    print("Not allowed, please use numbers only: ")
                elif (pattern == DATE_PATTERN):
                    print("Not allowed, please use the right format dd/mm/yyyy: ")
                answer = input()
        return answer

    # intakeNewDog method
    def intakeNewDog(self):
        name = input("What is the dog's name?\n");
        if(self.dataBaseServices.checkAnimalByNameType(name, "dog")):
            print("\n\nThis dog is already in our system\n\n")
            return # returns to menu

        breed = input("What is the dog's breed?\n")
        gender = input("What is the dog's gender?\n")
        print("What is the dog's age?")
        age = self.checkAnswer(DIGIT_PATTERN);
        print("What is the dog's weight in lbs?")
        weight = self.checkAnswer(DIGIT_PATTERN)
        print("What is the dog's acquisition date? In format: dd/mm/yyyy")
        acquisitionDate = self.checkAnswer(DATE_PATTERN)
        acquisitionCountry = input("What is the dog's acquisition country?\n")
        print("What is the dog's training status?\n"
                + "Possible answers: intake, farm, in-service, farm, Phase (I, II, III, IV, V)")
        # getting training status of a dog
        trainingStatus = input()
        # need it to stop the loop when we get the correct training status
        choice = True
        while(choice):
            if (trainingStatus not in trainingStatusAllowList):
                print("Whrong status! Please use the correct status!")
                trainingStatus = input()
            else:
                print("Thank you!\nAdding the dog in our database...")
                choice = False
        # default not reserved
        reserved = False
        # default acquisition country is equal to service country
        inServiceCountry = acquisitionCountry
        # adding a dog to our database
        self.dataBaseServices.addOneDogToDatabase(name, breed, gender, age, weight, acquisitionDate, acquisitionCountry, trainingStatus, reserved, inServiceCountry)
        print("Dog was added to our database.\nReturning to the main menu...");
        return


    # intakeNewMonkey
    def intakeNewMonkey(self):
        name = input("What is the monkey's name?\n")
        if(self.dataBaseServices.checkAnimalByNameType(name, "monkey")):
            print("\n\nThis dog is already in our system\n\n")
            return # returns to menu
        print("What is the monkey's species?\n"
                + "Currently supported species: Capuchin, Guenon, Macaque, Marmoset, Squirrel monkey, Tamarin")
        species = input()
        if (species not in allowedMonkeysSpeciesList):
            print("\n\nThis monkey species is not allowed\n\n")
            return
        gender = input("What is the monkey's gender?\n")
        print("What is the monkey's age?")
        age = self.checkAnswer(DIGIT_PATTERN)
        print("What is the monkey's weight in lbs?")
        weight = self.checkAnswer(DIGIT_PATTERN)
        print("What is the monkey's acquisition date? In format: dd/mm/yyyy")
        acquisitionDate = self.checkAnswer(DATE_PATTERN)
        acquisitionCountry = input("What is the monkey's acquisition country?\n")
        # default status is intake
        trainingStatus = "intake"
        # default is not reserved
        reserved = False
        # default acquisition country is equal to service country
        inServiceCountry = acquisitionCountry
        # continue asking for the rest of the required information such as monkeys height, bodyLength, tailLength
        print("What is the monkey's height in inches?")
        height = self.checkAnswer(DIGIT_PATTERN)

        print("What is the monkey's body length inches?")
        bodyLength = self.checkAnswer(DIGIT_PATTERN)

        print("What is the monkey's tail length inches?")
        tailLength = self.checkAnswer(DIGIT_PATTERN)

        # adding monkey to our database
        self.dataBaseServices.addOneMonkeyToDatabase(name, species, gender, age, weight, acquisitionDate, acquisitionCountry, trainingStatus, reserved, inServiceCountry, height, bodyLength, tailLength)
        print("Monkey was added. Returning to the main menu.");
        return


    # reserveAnimal method
    # We will need to find the animal by animal type and in service country
    def reserveAnimal(self):
        print("Please enter animal's type to be reserved?");
        type = input()
        choiceAnimal = True
        while(choiceAnimal):
            if ( (type != "dog") and (type != "monkey") ):
                print("Whrong type!\nPossible types: dog, monkey")
                type = input()
            else:
                choiceAnimal = False

        print("Please enter animal's country to be reserved?")
        inServiceCountry = input()
        result = self.dataBaseServices.reserveAnimalPerCountryAndType(inServiceCountry, type)
        if (result == -1):
            print("No animal to reserve in this country! Come back later!")
            return
        else:
            print(result + " is reserved!")
            return


    # method printAnimals
    # Includes the animal name, status, acquisition country and if the animal is reserved.
    # This method connects to three different menu items.
    # The printAnimals() method has three different outputs
    # based on the listType parameter
    # dog - prints the list of dogs
    # monkey - prints the list of monkeys
    # available - prints a combined list of all animals that are
    # fully trained ("in service") but not reserved
    def printAnimals(self, animalType):
        if (animalType == "monkey"):
            print("-------------------------------------------------------------------")
            print("Name" + (10 - len("Name"))*" ", "Training Status" + (15 - len("Training Status"))*" ", "Location"  + (10 - len("Type"))*" ", "Reserved", sep="  |  ")
            print("-------------------------------------------------------------------")
            self.dataBaseServices.printRescueAnimalsByType("monkey")
            print("-------------------------------------------------------------------\n")

        elif (animalType == "dog"):
            print("-------------------------------------------------------------------")
            print("Name" + (10 - len("Name"))*" ", "Training Status" + (15 - len("Training Status"))*" ", "Location"  + (10 - len("Type"))*" ", "Reserved", sep="  |  ")
            print("-------------------------------------------------------------------")
            self.dataBaseServices.printRescueAnimalsByType("dog")
            print("-------------------------------------------------------------------\n")

        else:
            print("--------------------------------------------------------------------------------")
            print("Name" + (10 - len("Name"))*" ", "Type" + (10 - len("Type"))*" ", "Training Status", "Location", "Reserved", sep="  |  ")
            print("--------------------------------------------------------------------------------")
            self.dataBaseServices.printReservedReady()

    # This method prints the menu options
    def displayMenu(self):
        print("\n\n");
        print("\t\t\t\tRescue Animal System Menu")
        print("[1] Intake a new dog")
        print("[2] Intake a new monkey")
        print("[3] Reserve an animal")
        print("[4] Print a table of all dogs")
        print("[5] Print a table of all monkeys")
        print("[6] Print a table of all animals that are ready to be reserved")
        print("[q] Quit application")
        print()
        print("Enter a menu selection")

    # main loop of choices
    def mainLoop(self):
        while(True):
            inputText = input("")
            match inputText:
                case "1":
                    self.intakeNewDog()
                case "2":
                    self.intakeNewMonkey()
                case "3":
                    self.reserveAnimal()
                case "4":
                    self.printAnimals("dog")
                case "5":
                    self.printAnimals("monkey")
                case "6":
                    self.printAnimals("")
                case "q":
                    print("\nSee you later! Data was preserved for future use")
                    return
                case _:
                    print("Please use numbers from 1 to 6, or q to exit")
            self.displayMenu()


    # our main function that checks username and password
    # if successfull we display our menu to perform further actions
    # if no user exists new one is created
    def startMain(self):
        # counters for bruteforce protection
        wrongUserCounter = 0
        wrongPasswordCounter = 0
        loggedOut = True
        while(loggedOut):
            print("----------Login Menu----------\n")
            username = input("Enter your username: ")
            results = self.dataBaseServices.getUserByName(username)
            if(results):
                password = maskpass.askpass("Enter your password: ")
                hashed_password = self.dataBaseServices.getPasswordSaltedHash(password)
                salt, key = results[0].saltedhash[:16], results[0].saltedhash[16:]
                hash_algo = "sha256"
                iterations = 100_000

                # get a hash from user entered password
                password_hash_check = self.dataBaseServices.getPBKDF2_hmac(hash_algo, password, salt, iterations)
                if (key == password_hash_check):
                    loggedOut = False
                    self.displayMenu()
                    self.mainLoop()
                else:
                    wrongPasswordCounter += 1
                    print("Wrong password!")
                    pass
            else:
                wrongUserCounter += 1
                print("User does not exist!")
            if (wrongUserCounter > 3 or wrongPasswordCounter > 3):
                # fix me
                # add 1 minutes wait time before trying again
                # store time in the database
                print("Too many attempts. Try again later in 1 minute!")
                time.sleep(60)
                wrongUserCounter = 0
                wrongPasswordCounter = 0
                # break

# Our Main Logic starts here
driver = Driver("RescueAnimal.db")

# import some data if our database is empty
if (driver.dataBaseServices.checkRescueAnimalsTableEmpty()):
    print("RescueAnimal is not empty, no need to initialize data.")
else:
    print("RescueAnimal is empty.\nInitializing data")
    driver.dataBaseServices.initializeData()
    print("Done")

# check if at least one user exists
# else create one
if (not driver.dataBaseServices.checkUsersTableEmpty()):
    print("User table is empty\nLet's create an Admin user\nMake sure you write it down somewhere and store it securely")
    username = input("Enter username: ")
    password = maskpass.askpass("Enter your password: ")
    driver.dataBaseServices.addUser(username, password)

# starting our Main function
driver.startMain()
